// sillytavern-utils-lib/dist/types/index.js
export var EventNames;
(function (EventNames) {
    EventNames["APP_READY"] = "app_ready";
    EventNames["EXTRAS_CONNECTED"] = "extras_connected";
    EventNames["MESSAGE_SWIPED"] = "message_swiped";
    EventNames["MESSAGE_SENT"] = "message_sent";
    EventNames["MESSAGE_RECEIVED"] = "message_received";
    EventNames["MESSAGE_EDITED"] = "message_edited";
    EventNames["MESSAGE_DELETED"] = "message_deleted";
    EventNames["MESSAGE_UPDATED"] = "message_updated";
    EventNames["MESSAGE_FILE_EMBEDDED"] = "message_file_embedded";
    EventNames["MORE_MESSAGES_LOADED"] = "more_messages_loaded";
    EventNames["IMPERSONATE_READY"] = "impersonate_ready";
    EventNames["CHAT_CHANGED"] = "chat_id_changed";
    EventNames["GENERATION_AFTER_COMMANDS"] = "GENERATION_AFTER_COMMANDS";
    EventNames["GENERATION_STARTED"] = "generation_started";
    EventNames["GENERATION_STOPPED"] = "generation_stopped";
    EventNames["GENERATION_ENDED"] = "generation_ended";
    EventNames["EXTENSIONS_FIRST_LOAD"] = "extensions_first_load";
    EventNames["EXTENSION_SETTINGS_LOADED"] = "extension_settings_loaded";
    EventNames["SETTINGS_LOADED"] = "settings_loaded";
    EventNames["SETTINGS_UPDATED"] = "settings_updated";
    EventNames["GROUP_UPDATED"] = "group_updated";
    EventNames["MOVABLE_PANELS_RESET"] = "movable_panels_reset";
    EventNames["SETTINGS_LOADED_BEFORE"] = "settings_loaded_before";
    EventNames["SETTINGS_LOADED_AFTER"] = "settings_loaded_after";
    EventNames["CHATCOMPLETION_SOURCE_CHANGED"] = "chatcompletion_source_changed";
    EventNames["CHATCOMPLETION_MODEL_CHANGED"] = "chatcompletion_model_changed";
    EventNames["OAI_PRESET_CHANGED_BEFORE"] = "oai_preset_changed_before";
    EventNames["OAI_PRESET_CHANGED_AFTER"] = "oai_preset_changed_after";
    EventNames["OAI_PRESET_EXPORT_READY"] = "oai_preset_export_ready";
    EventNames["OAI_PRESET_IMPORT_READY"] = "oai_preset_import_ready";
    EventNames["WORLDINFO_SETTINGS_UPDATED"] = "worldinfo_settings_updated";
    EventNames["WORLDINFO_UPDATED"] = "worldinfo_updated";
    EventNames["CHARACTER_EDITED"] = "character_edited";
    EventNames["CHARACTER_PAGE_LOADED"] = "character_page_loaded";
    EventNames["CHARACTER_GROUP_OVERLAY_STATE_CHANGE_BEFORE"] = "character_group_overlay_state_change_before";
    EventNames["CHARACTER_GROUP_OVERLAY_STATE_CHANGE_AFTER"] = "character_group_overlay_state_change_after";
    EventNames["USER_MESSAGE_RENDERED"] = "user_message_rendered";
    EventNames["CHARACTER_MESSAGE_RENDERED"] = "character_message_rendered";
    EventNames["FORCE_SET_BACKGROUND"] = "force_set_background";
    EventNames["CHAT_DELETED"] = "chat_deleted";
    EventNames["CHAT_CREATED"] = "chat_created";
    EventNames["GROUP_CHAT_DELETED"] = "group_chat_deleted";
    EventNames["GROUP_CHAT_CREATED"] = "group_chat_created";
    EventNames["GENERATE_BEFORE_COMBINE_PROMPTS"] = "generate_before_combine_prompts";
    EventNames["GENERATE_AFTER_COMBINE_PROMPTS"] = "generate_after_combine_prompts";
    EventNames["GENERATE_AFTER_DATA"] = "generate_after_data";
    EventNames["GROUP_MEMBER_DRAFTED"] = "group_member_drafted";
    EventNames["GROUP_WRAPPER_STARTED"] = "group_wrapper_started";
    EventNames["GROUP_WRAPPER_FINISHED"] = "group_wrapper_finished";
    EventNames["WORLD_INFO_ACTIVATED"] = "world_info_activated";
    EventNames["TEXT_COMPLETION_SETTINGS_READY"] = "text_completion_settings_ready";
    EventNames["CHAT_COMPLETION_SETTINGS_READY"] = "chat_completion_settings_ready";
    EventNames["CHAT_COMPLETION_PROMPT_READY"] = "chat_completion_prompt_ready";
    EventNames["CHARACTER_FIRST_MESSAGE_SELECTED"] = "character_first_message_selected";
    // TODO: Naming convention is inconsistent with other events
    EventNames["CHARACTER_DELETED"] = "characterDeleted";
    EventNames["CHARACTER_DUPLICATED"] = "character_duplicated";
    EventNames["CHARACTER_RENAMED"] = "character_renamed";
    /** @deprecated The event is aliased to STREAM_TOKEN_RECEIVED. */
    EventNames["SMOOTH_STREAM_TOKEN_RECEIVED"] = "stream_token_received";
    EventNames["STREAM_TOKEN_RECEIVED"] = "stream_token_received";
    EventNames["STREAM_REASONING_DONE"] = "stream_reasoning_done";
    EventNames["FILE_ATTACHMENT_DELETED"] = "file_attachment_deleted";
    EventNames["WORLDINFO_FORCE_ACTIVATE"] = "worldinfo_force_activate";
    EventNames["OPEN_CHARACTER_LIBRARY"] = "open_character_library";
    EventNames["ONLINE_STATUS_CHANGED"] = "online_status_changed";
    EventNames["IMAGE_SWIPED"] = "image_swiped";
    EventNames["CONNECTION_PROFILE_LOADED"] = "connection_profile_loaded";
    EventNames["TOOL_CALLS_PERFORMED"] = "tool_calls_performed";
    EventNames["TOOL_CALLS_RENDERED"] = "tool_calls_rendered";
})(EventNames || (EventNames = {}));
